File: bench/node/ring.js
// A ring of N async nodes; a token travels around +1/hop for LAPS laps
// (N*LAPS hops). Node's single-threaded event loop is its concurrency model.
// Checksum = N*LAPS.
const N = parseInt(process.env.BENCH_N || "200", 10);
const LAPS = 5000;
const total = N * LAPS;

class Chan {
  constructor() { this.q = []; this.w = []; }
  send(v) { const r = this.w.shift(); if (r) r(v); else this.q.push(v); }
  recv() { return new Promise((res) => { if (this.q.length) res(this.q.shift()); else this.w.push(res); }); }
}

const chans = Array.from({ length: N }, () => new Chan());
let resolveDone;
const donePromise = new Promise((r) => (resolveDone = r));
for (let i = 0; i < N; i++) {
  const inbox = chans[i], next = chans[(i + 1) % N];
  (async () => {
    while (true) {
      const v = await inbox.recv();
      if (v >= total) { resolveDone(v); break; }
      next.send(v + 1);
    }
  })();
}
chans[0].send(0);
donePromise.then((v) => console.log(v));

File: bench/go/ring.go
// A ring of N goroutines; a token travels around +1/hop for LAPS laps (N*LAPS hops).
// Checksum = N*LAPS.
package main

import "fmt"

func main() {
	n := benchN(200)
	const laps = 5000
	total := n * laps
	chans := make([]chan int, n)
	for i := range chans {
		chans[i] = make(chan int, 1)
	}
	done := make(chan int)
	for i := 0; i < n; i++ {
		inbox, next := chans[i], chans[(i+1)%n]
		go func() {
			for v := range inbox {
				if v >= total {
					done <- v
					return
				}
				next <- v + 1
			}
		}()
	}
	chans[0] <- 0
	fmt.Println(<-done)
}
